`sector_profile()` now accounts for unmatched `type`, `sector` or `subsector`

maurolepore opened the pull request:
* Closes #638 
* Relates to #639 
* Note to self: https://github.com/2DegreesInvesting/tiltIndicator/discussions/782

--

Given a `clustered` matching one but not a second `type` of scenario, when the `scenarios` dataset has the two types, then the second `type` and its corresponding `scenario` are still present in `grouped_by`, and the mismatch is reflected correctly in the `value`.

The expected behaviour is captured in [this GoogleSheet](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064) and explained in https://github.com/2DegreesInvesting/tiltIndicator/pull/739#issuecomment-1977426095 (thanks @Tilmon).

----

TODO

- [x] [Link related issue/PR]([url](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)). 
- [ ] Describe the goal of the PR. Avoid details that are clear in the diff.
- [x] Mark the PR as draft.
- [x] [Include a unit test](https://code-review.tidyverse.org/reviewer/aspects.html#sec-tests).
- [ ] Review your own PR in "Files changed".
- [ ] Ensure the PR branch is updated.
- [ ] Ensure the checks pass.
- [ ] Change the status from draft to ready.
- [ ] Polish the PR title and description.
- [ ] Assign a reviewer.

EXCEPTIONS

- [ ] Slide here any item that you intentionally choose to not do.

maurolepore commented:
I'm struggling to see the difference between the case "unmatched products" and the case "missing benchmarks".

For `emissions*()` I could clearly understand the difference and and create different tests for the cases with "unmatched product" versus "missing benchmark". For example, to test the case "unmatched product" I could add a product in the `companies` dataset that did not exist in the `products` dataset. And to test the case "missing benchmark" I could add an `NA` in the `isic_4digit` column of the `products` dataset.

However, for `sector*()` I can't understand the difference as clearly, and I realize that my tests for both cases essentially use the same type of toy input data: An `NA` or "unmatched" in either the columns `sector`, `subsector`, or `type` of the `companies` dataset. 

Your https://github.com/2DegreesInvesting/tiltIndicator/pull/738#issuecomment-1975997275 seems to support the idea that these two cases are no different:

> I agree that in this case the unmatched value in sector is the important relationship.

The best way to clarify things is through a reproducible example that you are familiar with because it's based on the one you created in this [GoogleSheet](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064). At this point I'm so confused that I don't know if you think it shows one or both cases, but I hope it's a good start.

<details/>
<summary/>
reprex
</summary>

``` r
library(tibble)
devtools::load_all()
#> ℹ Loading tiltIndicator

packageVersion("tiltIndicator")
#> [1] '0.0.0.9211'

withr::local_options(list(tibble.print_max = Inf, width = 500))

companies <- tribble(
  ~companies_id, ~clustered, ~activity_uuid_product_uuid, ~tilt_sector, ~tilt_subsector,       ~type,     ~sector,  ~subsector,
            "a",        "a",                         "a",          "a",             "a",       "ipr",     "total",    "energy",
            "a",        "a",                         "a",          "a",             "a",       "weo",     "total",    "energy",
            "a",        "b",                 "unmatched",  "unmatched",     "unmatched", "unmatched", "unmatched", "unmatched",
            "a",        "c",                 "unmatched",          "c",             "c",       "ipr", "land use",   "land use",
            "a",        "c",                 "unmatched",          "c",             "c",       "weo",         NA,           NA
  )

scenarios <- tribble(
     ~sector,   ~subsector,  ~year, ~reductions, ~type, ~scenario,
     "total",     "energy",   2050,         1.0, "ipr",       "a",
     "total",     "energy",   2050,         0.6, "weo",       "a",
  "land use",   "land use",   2050,         0.3, "ipr",       "a"
)

sector_profile(companies, scenarios) |> unnest_product()
#> # A tibble: 5 × 11
#>   companies_id grouped_by risk_category profile_ranking clustered activity_uuid_product_uuid tilt_sector scenario  year type      tilt_subsector
#>   <chr>        <chr>      <chr>                   <dbl> <chr>     <chr>                      <chr>       <chr>    <dbl> <chr>     <chr>         
#> 1 a            ipr_a_2050 high                      1   a         a                          a           a         2050 ipr       a             
#> 2 a            weo_a_2050 medium                    0.6 a         a                          a           a         2050 weo       a             
#> 3 a            <NA>       <NA>                     NA   b         unmatched                  unmatched   <NA>        NA unmatched unmatched     
#> 4 a            ipr_a_2050 low                       0.3 c         unmatched                  c           a         2050 ipr       c             
#> 5 a            <NA>       <NA>                     NA   c         unmatched                  c           <NA>        NA weo       c

sector_profile(companies, scenarios) |> unnest_company()
#> # A tibble: 8 × 4
#>   companies_id grouped_by risk_category value
#>   <chr>        <chr>      <chr>         <dbl>
#> 1 a            ipr_a_2050 high          0.25 
#> 2 a            ipr_a_2050 medium        0    
#> 3 a            ipr_a_2050 low           0.25 
#> 4 a            ipr_a_2050 <NA>          0.5  
#> 5 a            weo_a_2050 high          0    
#> 6 a            weo_a_2050 medium        0.333
#> 7 a            weo_a_2050 low           0    
#> 8 a            weo_a_2050 <NA>          0.667
```

</details>

It it possible that this is already all you want? If yes, we're done and I can close this PR. If not, what do you expect?

Tilmon commented:
Hi @maurolepore , 

let me respond to the points you raised below. I first start with a more detailed description of how to think about the two different cases and then explain why I think the reprex is unfortunately wrong. 

**Description of the "two cases"**
> I'm struggling to see the difference between the case "unmatched products" and the case "missing benchmarks".

I must admit that I find the difference in the case of `sector*()` also more complicated, but it's still there. Let's think about it that way: For the `sector*()` it doesn't matter, whether a `clustered `is matched to Ecoinvent or not. What matters is:

1. Whether the `clustered `has a `tilt_sector`. If it doesn't have a `tilt_sector `, the `clustered ` will neither have a `sector `or `subsector `for any of the `types`. Hence, you can think of the `tilt_sector` as the `activity_uuid_product_uuid` of `sector*()`: If the `tilt_sector ` (`activity_uuid_product_uuid`) is missing, you won't have any `sector` data (`co2_footprint`) that can be used to calculate the profile. Because ultimately, the scenario sectors give us the info on the reduction targets. I.e., no scenario sector = no result at all. Hence, `tilt_sector `== `unmatched `should be have as the `activity_uuid_product_uuid `== unmatched in the emission*(). 
2. If the `tilt_sector` & `tilt_subsector ` leads to a `sector `& `subsector ` for either of the `type` ipr or weo or both or none. You can think of the `sector `x `subsector `x `type `x `year` combination as the 6 benchmarks in the `emission*()` . For each `clustered `with a `tilt_sector`, we want to show all benchmarks, i.e. all corresponding combinations of `sector `x `subsector `x `type `x `year` for the specific `tilt_sector `x `tilt_subsector `, even if some are NAs. I.e., for every `clustered `with a `tilt_sector`, we should show the benchmarks weo_2030, weo_2050, ipr_2030, ipr_2050, even if some are NA (as in your reprex `clustered `c has no corresponding weo sector) - similar to an NA in `isic_4digit `in `emission_profile()`.

**Reprex**
Your reprex shows BOTH cases. 
- `clustered` b has no `tilt_sector` (equivalent of no `activity_uuid_product_uuid`) and will hence lead to no results for the indicator (as there is no `sector`, equivalent to no `co2_footprint`). This is what I describe under 1. 
- `clustered` c has a `tilt_sector` (equivalent to matched product with `activity_uuid_product_uuid`) and hence will lead to results for the `sector `x `subsector `x `type `x `year `combination, even if there will be some NAs (equivalent to an `activity_uuid_product_uuid` leading to results for at least some of the 6 benchmarks). 

So the reprex example is great to discuss the issue. I see two problems with the reprex you shared, one on product-level and one on company-level: 
- **product-level:** clustered c should have the `grouped_by `value "weo_a_2050" instead of "NA". We have a `tilt_sector` for that clustered and hence should show all benchmarks, even if they are NA. 
- **company-level:** The results show that for ipr_a_2050, there are 25% of products in high and 25% in low risk category. That can't be right, because we only have 3 products. Instead, we have 1 clustered with high, 1 with low, and one with NA benchmark , i.e. we should have for the benchmark ipr_a_2050 the following values: high 1/3, medium 0, low 1/3, NA 1/3. 


P.S. I realize my comment here was not very helpful, as it's not wrong but leads to more confusion about the two different cases. 

> Your https://github.com/2DegreesInvesting/tiltIndicator/pull/738#issuecomment-1975997275 seems to support the idea that these two cases are no different: "But I agree that in this case the unmatched value in sector is the important relationship."

I agree that in this case the unmatched value in sector is the important relationship.

maurolepore commented on tests/testthat/test-sector_profile.R:
```diff
@@ -127,3 +127,30 @@ test_that("at company level, two matched and one unmatched products yield `value
   other <- pull(filter(out, !is.na(risk_category)), value)
   expect_equal(sort(other), c(0, 0, 2 / 3))
 })
+
+test_that("at product level, when `companies$tilt_sector` doesn't match `scenarios$sector` then `product$grouped_by` is `NA`", {
+  # This is what I think an analyst would say:
+  # > A product's tilt_sector with unmatched scenarios's sector yields a "missing benchmark"
+  #
+  # Translation to tiltIndicator's code:
+  # * given: the value "b" of `companies$clustered`
+  # * when: the corresponding value of `companies$tilt_sector` doesn't match any value of `scenarios$sector`
+  # * then: the result at product level has the value `NA` in `product$grouped_by`
+
+  # Based on this GoogleSheet row:
+  # https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064&range=A5:I5
+  # styler: off
+  companies <- tribble(
+    ~companies_id, ~clustered, ~activity_uuid_product_uuid, ~tilt_sector, ~tilt_subsector,       ~type,     ~sector,  ~subsector,
+              "a",        "b",                 "unmatched",  "unmatched",     "unmatched", "unmatched", "unmatched", "unmatched",
+  )
+  scenarios <- tribble(
+    ~sector,   ~subsector,  ~year, ~reductions, ~type, ~scenario,
+    "total",     "energy",   2050,         1.0, "ipr",       "a",
+  )
+  # styler: on
+
+  product <- sector_profile(companies, scenarios) |> unnest_product()
+
+  expect_true(is.na(product$grouped_by))
+})
```
@Tilmon, 

Below I unpack a piece of your https://github.com/2DegreesInvesting/tiltIndicator/pull/739#issuecomment-1977426095. 

I turned my understanding into a test (highlighted code), and it's corresponding reprex for your convenience. Please review and correct any misunderstanding.

<details/>
<summary/>
reprex
</summary>

``` r
devtools::load_all()
#> ℹ Loading tiltIndicator

companies <- tribble(
  ~companies_id, ~clustered, ~activity_uuid_product_uuid, ~tilt_sector, ~tilt_subsector,       ~type,     ~sector,  ~subsector,
            "a",        "b",                 "unmatched",  "unmatched",     "unmatched", "unmatched", "unmatched", "unmatched",
)
scenarios <- tribble(
  ~sector,   ~subsector,  ~year, ~reductions, ~type, ~scenario,
  "total",     "energy",   2050,         1.0, "ipr",       "a",
)
# styler: on

product <- sector_profile(companies, scenarios) |> unnest_product()
product
#> # A tibble: 1 × 11
#>   companies_id grouped_by risk_category profile_ranking clustered
#>   <chr>        <chr>      <chr>                   <dbl> <chr>    
#> 1 a            <NA>       <NA>                       NA b        
#> # ℹ 6 more variables: activity_uuid_product_uuid <chr>, tilt_sector <chr>,
#> #   scenario <chr>, year <dbl>, type <chr>, tilt_subsector <chr>

is.na(product$grouped_by)
#> [1] TRUE
```

</details>

----

> For the `sector*()` ... what matters is:

> Whether the clustered has a tilt_sector.

When constructing toy data focus on `companies$tilt_sector`, then adjust `companies$sector` accordingly (see next).

> [If it doesn't have a tilt_sector]^1, the clustered [will neither have a sector or subsector for any of the types]^2.

* ^1: Means {if the value of `companies$tilt_sector` doesn't match any
`scenarios$sector`} ([example](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064&range=A5:I5)).

* ^2: Means {yields `product$grouped_by == NA`} ([example](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064&range=B19)).

Clarifications:

> If the tilt_sector ... is [missing]^3.

* ^3: Here, "missing" does not mean `NA` neither it means the absence of a row. What this means is ^1 above.

> no scenario sector = [no result at all]^4.

* ^4: Here, "no result at all" doesn't mean no output. It means ^2.

> clustered b [has no tilt_sector]^5 ... and will hence lead to [no results]^6

* ^5: Here, "has no tilt_sector" doesn't mean `NA` or absence. It means ^1.

* ^6: Here, "no results" doesn't mean no output. It means ^2.

maurolepore commented on tests/testthat/test-sector_profile.R:
```diff
@@ -127,3 +127,51 @@ test_that("at company level, two matched and one unmatched products yield `value
+  expect_true(is.na(product$grouped_by))
+})
+
+test_that("at product level, when `companies$tilt_sector` doesn't match `scenarios$sector` then `product$risk_category` and  `product$profile_ranking` are `NA`", {
+  # TODO: DRY with the test above
+  # styler: off
+  companies <- tribble(
+    ~companies_id, ~clustered, ~activity_uuid_product_uuid, ~tilt_sector, ~tilt_subsector,       ~type,     ~sector,  ~subsector,
+              "a",        "b",                 "unmatched",  "unmatched",     "unmatched", "unmatched", "unmatched", "unmatched",
+  )
+  scenarios <- tribble(
+    ~sector,   ~subsector,  ~year, ~reductions, ~type, ~scenario,
+    "total",     "energy",   2050,         1.0, "ipr",       "a",
+  )
+  # styler: on
+
+  product <- sector_profile(companies, scenarios) |> unnest_product()
+
+  # Tilman's expectations
+  # https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064&range=B19:D19
+  expect_true(is.na(product$risk_category))
+  expect_true(is.na(product$profile_ranking))
+})
```
@Tilmon, 

This test is exactly like the one above but focuses on other expectations that you didn't mention in https://github.com/2DegreesInvesting/tiltIndicator/pull/739#issuecomment-1977426095 but are implicit from your [GoogleSheet](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064). 

Later I may consolidate these expectations with the ones above, but for now they are a useful placeholder to discuss other results I didn't include in this test yet. 

Specifically, some values in the result differ from what you show in your GoogleSheet. Please see this reprex and let me know if I need to fix the code or if the current output is OK.

<details/>
<summary/>
reprex
</summary>

``` r
if (!interactive()) withr::local_options(width = 500)
devtools::load_all()
#> ℹ Loading tiltIndicator

companies <- tribble(
  ~companies_id, ~clustered, ~activity_uuid_product_uuid, ~tilt_sector, ~tilt_subsector,       ~type,     ~sector,  ~subsector,
            "a",        "b",                 "unmatched",  "unmatched",     "unmatched", "unmatched", "unmatched", "unmatched",
)
scenarios <- tribble(
  ~sector,   ~subsector,  ~year, ~reductions, ~type, ~scenario,
  "total",     "energy",   2050,         1.0, "ipr",       "a",
)
# styler: on

product <- sector_profile(companies, scenarios) |> unnest_product()
product
#> # A tibble: 1 × 11
#>   companies_id grouped_by risk_category profile_ranking clustered activity_uuid_product_uuid tilt_sector scenario  year type      tilt_subsector
#>   <chr>        <chr>      <chr>                   <dbl> <chr>     <chr>                      <chr>       <chr>    <dbl> <chr>     <chr>         
#> 1 a            <NA>       <NA>                       NA b         unmatched                  unmatched   <NA>        NA unmatched unmatched

# Different from the GoogleSheet
different <- c("scenario", "type", "tilt_subsector")
product |> select(all_of(different))
#> # A tibble: 1 × 3
#>   scenario type      tilt_subsector
#>   <chr>    <chr>     <chr>         
#> 1 <NA>     unmatched unmatched
```

</details>





For your convenience, here is a screenshot. The relevant row is the one highlighted.

![image](https://github.com/2DegreesInvesting/tiltIndicator/assets/5856545/1a6c4404-e612-4f7d-a41b-dd2993070c2d)

maurolepore commented on tests/testthat/test-sector_profile.R:
```diff
@@ -127,3 +127,71 @@ test_that("at company level, two matched and one unmatched products yield `value
+  expect_true(is.na(product$profile_ranking))
+})
+
+test_that("at product level, when `companies$sector`, `companies$subsector`, and `companies$type` match the corresponding columns in `scenarios`, then  `product$grouped_by` is '<type>_<scenario>_<year>' and `product$risk_category` and `product$profile_ranking` are not `NA`", {
+  # styler: off
+  companies <- tribble(
+    ~companies_id, ~clustered, ~activity_uuid_product_uuid, ~tilt_sector, ~tilt_subsector,       ~type,     ~sector,  ~subsector,
+              "a",        "a",                         "a",          "a",             "a",       "ipr",     "total",    "energy",
+  )
+  # matching type, sector and subsector
+  scenarios <- tribble(
+    ~sector,   ~subsector,  ~year, ~reductions, ~type, ~scenario,
+    "total",     "energy",   2050,         1.0, "ipr",       "a",
+  )
+  # styler: on
+
+  product <- sector_profile(companies, scenarios) |> unnest_product()
+
+  expect_equal(product$grouped_by, "ipr_a_2050")
+  expect_false(is.na(product$risk_category))
+  expect_false(is.na(product$profile_ranking))
+})
```
This is the “happy path” at product level:

- Given: a value of `companies$clustered`, `companies$tilt_sector`, `companies$tilt_sector`, and `companies$type` ([example](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064&range=B3:F3))
- Where: the value of `companies$type` matches `scenarios$type`, and the values of `companies$sector` and `companies$subsector` ([example](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064&range=G3:H3)) matches [the values of `scenarios$sector` and `scenarios$subsector`](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064&range=A11:B11))
- Then: the value of `product$grouped_by` is`‘<type>*<scenario>*<year>’`([example](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064&range=B17)), and the value of `product$risk_cagegory` and `product$profile_ranking` are not `NA`.

<details/>
<summary/>
reprex
</summary>

``` r
withr::local_options(width = 500)
devtools::load_all()
#> ℹ Loading tiltIndicator

# styler: off
companies <- tribble(
  ~companies_id, ~clustered, ~activity_uuid_product_uuid, ~tilt_sector, ~tilt_subsector,       ~type,     ~sector,  ~subsector,
            "a",        "a",                         "a",          "a",             "a",       "ipr",     "total",    "energy",
)
# Matching type, sector and subsector
scenarios <- tribble(
  ~sector,   ~subsector,  ~year, ~reductions, ~type, ~scenario,
  "total",     "energy",   2050,         1.0, "ipr",       "a",
)
# styler: on

product <- sector_profile(companies, scenarios) |> unnest_product()
product
#> # A tibble: 1 × 11
#>   companies_id grouped_by risk_category profile_ranking clustered activity_uuid_product_uuid tilt_sector scenario  year type  tilt_subsector
#>   <chr>        <chr>      <chr>                   <dbl> <chr>     <chr>                      <chr>       <chr>    <dbl> <chr> <chr>         
#> 1 a            ipr_a_2050 high                        1 a         a                          a           a         2050 ipr   a
```

</details>

maurolepore commented:
@Tilmon 

RE:

> product-level: clustered c should have the grouped_by value "weo_a_2050" instead of "NA". 

Can you please confirm you expect 1, 2, 3, or something else?

1. `"weo_a_2050"`
2. `"weo_NA_2050"`
3. `"weo_NA_NA"`

To me "1." seems incorrect. The values of `grouped_by` have the format `<type>_<scenario>_<year>`. And a `scenario`  and `year` that exist for a given `type` and `year` may not make sense for another `type`. 

For example, if a `clustered` "d" matches the `sector` and `subsector` for the `type` "ipr", and the corresponding `scenario` is called "iprScenario2050", then I expect `grouped_by == "ipr_iprScenario2050_2050` -- as you say. But if that same `sector` and `subsector` is unmatched for the `type` "weo", then expecting `grouped_by == "weo_iprScenario2050_2050"` seems very odd.

maurolepore commented:
@Tilmon, 

Here I reproduce the example from your [GoogleSheet](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064), first case by case then all at once. Please review and try to explain how we can change this output to meet your expectations. 

Case by case the outputs seem to make sense but when taken together the output clearly has more values of `grouped_by` than the ones you expect. The solution may be to gather consolidate the rows where `grouped_by` is `unmatched_NA_NA` and `weo_NA_NA`. But now I need a sleeping-break so I pass the thinking ball to you :-)

<details/>
<summary/>
reprex
</summary>

``` r
devtools::load_all()
#> ℹ Loading tiltIndicator

companies <- tribble(
  ~companies_id, ~clustered, ~activity_uuid_product_uuid, ~tilt_sector, ~tilt_subsector,       ~type,     ~sector,  ~subsector,
            "a",        "a",                         "a",          "a",             "a",       "ipr",     "total",    "energy",
            "a",        "a",                         "a",          "a",             "a",       "weo",     "total",    "energy",
            "a",        "b",                 "unmatched",  "unmatched",     "unmatched", "unmatched", "unmatched", "unmatched",
            "a",        "c",                 "unmatched",          "c",             "c",       "ipr", "land use",   "land use",
            "a",        "c",                 "unmatched",          "c",             "c",       "weo",         NA,           NA
)

scenarios <- tribble(
     ~sector,   ~subsector,  ~year, ~reductions, ~type, ~scenario,
     "total",     "energy",   2050,         1.0, "ipr",       "a",
     "total",     "energy",   2050,         0.6, "weo",       "a",
  "land use",   "land use",   2050,         0.3, "ipr",       "a"
)

# CASE BY CASE ---------------------------------------------------------------

case_a <- filter(companies, clustered == "a")
case_a
#> # A tibble: 2 × 8
#>   companies_id clustered activity_uuid_produc…¹ tilt_sector tilt_subsector type 
#>   <chr>        <chr>     <chr>                  <chr>       <chr>          <chr>
#> 1 a            a         a                      a           a              ipr  
#> 2 a            a         a                      a           a              weo  
#> # ℹ abbreviated name: ¹​activity_uuid_product_uuid
#> # ℹ 2 more variables: sector <chr>, subsector <chr>

sector_profile(case_a, scenarios) |> unnest_product()
#> # A tibble: 2 × 11
#>   companies_id grouped_by risk_category profile_ranking clustered
#>   <chr>        <chr>      <chr>                   <dbl> <chr>    
#> 1 a            ipr_a_2050 high                      1   a        
#> 2 a            weo_a_2050 medium                    0.6 a        
#> # ℹ 6 more variables: activity_uuid_product_uuid <chr>, tilt_sector <chr>,
#> #   scenario <chr>, year <dbl>, type <chr>, tilt_subsector <chr>

sector_profile(case_a, scenarios) |> unnest_company()
#> # A tibble: 8 × 4
#>   companies_id grouped_by risk_category value
#>   <chr>        <chr>      <chr>         <dbl>
#> 1 a            ipr_a_2050 high              1
#> 2 a            ipr_a_2050 medium            0
#> 3 a            ipr_a_2050 low               0
#> 4 a            ipr_a_2050 <NA>              0
#> 5 a            weo_a_2050 high              0
#> 6 a            weo_a_2050 medium            1
#> 7 a            weo_a_2050 low               0
#> 8 a            weo_a_2050 <NA>              0




case_b <- filter(companies, clustered == "b")
case_b
#> # A tibble: 1 × 8
#>   companies_id clustered activity_uuid_produc…¹ tilt_sector tilt_subsector type 
#>   <chr>        <chr>     <chr>                  <chr>       <chr>          <chr>
#> 1 a            b         unmatched              unmatched   unmatched      unma…
#> # ℹ abbreviated name: ¹​activity_uuid_product_uuid
#> # ℹ 2 more variables: sector <chr>, subsector <chr>

sector_profile(case_b, scenarios) |> unnest_product()
#> # A tibble: 1 × 11
#>   companies_id grouped_by      risk_category profile_ranking clustered
#>   <chr>        <chr>           <chr>                   <dbl> <chr>    
#> 1 a            unmatched_NA_NA <NA>                       NA b        
#> # ℹ 6 more variables: activity_uuid_product_uuid <chr>, tilt_sector <chr>,
#> #   scenario <chr>, year <dbl>, type <chr>, tilt_subsector <chr>

sector_profile(case_b, scenarios) |> unnest_company()
#> # A tibble: 4 × 4
#>   companies_id grouped_by      risk_category value
#>   <chr>        <chr>           <chr>         <dbl>
#> 1 a            unmatched_NA_NA high              0
#> 2 a            unmatched_NA_NA medium            0
#> 3 a            unmatched_NA_NA low               0
#> 4 a            unmatched_NA_NA <NA>              1




case_c <- filter(companies, clustered == "c")
case_c
#> # A tibble: 2 × 8
#>   companies_id clustered activity_uuid_produc…¹ tilt_sector tilt_subsector type 
#>   <chr>        <chr>     <chr>                  <chr>       <chr>          <chr>
#> 1 a            c         unmatched              c           c              ipr  
#> 2 a            c         unmatched              c           c              weo  
#> # ℹ abbreviated name: ¹​activity_uuid_product_uuid
#> # ℹ 2 more variables: sector <chr>, subsector <chr>

sector_profile(case_c, scenarios) |> unnest_product()
#> # A tibble: 2 × 11
#>   companies_id grouped_by risk_category profile_ranking clustered
#>   <chr>        <chr>      <chr>                   <dbl> <chr>    
#> 1 a            ipr_a_2050 low                       0.3 c        
#> 2 a            weo_NA_NA  <NA>                     NA   c        
#> # ℹ 6 more variables: activity_uuid_product_uuid <chr>, tilt_sector <chr>,
#> #   scenario <chr>, year <dbl>, type <chr>, tilt_subsector <chr>

sector_profile(case_c, scenarios) |> unnest_company()
#> # A tibble: 8 × 4
#>   companies_id grouped_by risk_category value
#>   <chr>        <chr>      <chr>         <dbl>
#> 1 a            ipr_a_2050 high              0
#> 2 a            ipr_a_2050 medium            0
#> 3 a            ipr_a_2050 low               1
#> 4 a            ipr_a_2050 <NA>              0
#> 5 a            weo_NA_NA  high              0
#> 6 a            weo_NA_NA  medium            0
#> 7 a            weo_NA_NA  low               0
#> 8 a            weo_NA_NA  <NA>              1




# ALL AT ONCE ----------------------------------------------------------------

sector_profile(companies, scenarios) |> unnest_product()
#> # A tibble: 5 × 11
#>   companies_id grouped_by      risk_category profile_ranking clustered
#>   <chr>        <chr>           <chr>                   <dbl> <chr>    
#> 1 a            ipr_a_2050      high                      1   a        
#> 2 a            weo_a_2050      medium                    0.6 a        
#> 3 a            unmatched_NA_NA <NA>                     NA   b        
#> 4 a            ipr_a_2050      low                       0.3 c        
#> 5 a            weo_NA_NA       <NA>                     NA   c        
#> # ℹ 6 more variables: activity_uuid_product_uuid <chr>, tilt_sector <chr>,
#> #   scenario <chr>, year <dbl>, type <chr>, tilt_subsector <chr>

sector_profile(companies, scenarios) |> unnest_company()
#> # A tibble: 16 × 4
#>    companies_id grouped_by      risk_category value
#>    <chr>        <chr>           <chr>         <dbl>
#>  1 a            ipr_a_2050      high            0.5
#>  2 a            ipr_a_2050      medium          0  
#>  3 a            ipr_a_2050      low             0.5
#>  4 a            ipr_a_2050      <NA>            0  
#>  5 a            unmatched_NA_NA high            0  
#>  6 a            unmatched_NA_NA medium          0  
#>  7 a            unmatched_NA_NA low             0  
#>  8 a            unmatched_NA_NA <NA>            1  
#>  9 a            weo_NA_NA       high            0  
#> 10 a            weo_NA_NA       medium          0  
#> 11 a            weo_NA_NA       low             0  
#> 12 a            weo_NA_NA       <NA>            1  
#> 13 a            weo_a_2050      high            0  
#> 14 a            weo_a_2050      medium          1  
#> 15 a            weo_a_2050      low             0  
#> 16 a            weo_a_2050      <NA>            0
```

</details>

Tilmon commented:
Hi @maurolepore, thanks for your thorough explanations and for taking the time to reproduce the example from the [Google Sheet](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064) case-by-case and all-at-once. That really helps me to understand the struggles in replicating it. And it also shows how incredibly complicated it is. Therefore, I'd like to ask @AnneSchoenauer to also read through these comments carefully to validate my thinking. Maybe I'm also making things too complicated and we need to find a shortcut. 

Before I get into the details, one more general questions. I see in the outputs the column `profile_ranking`. Does this column show the `reductions`? Then can we also call it that way? Or does it show something different I'm unaware of? 

Now to your first question in https://github.com/2DegreesInvesting/tiltIndicator/pull/739#issuecomment-2110919379

> Can you please confirm you expect 1, 2, 3, or something else?
> 
> "weo_a_2050"
> "weo_NA_2050"
> "weo_NA_NA"

I would still say, the `grouped_by `should be "weo_a_2050" for `clustered `c. If we know the `type `for a `clustered`, then we can also assign a `scenario `and `year`, no? The scenarios table you are using in  https://github.com/2DegreesInvesting/tiltIndicator/pull/739#issuecomment-2111332672, contains the type-scenario-year combination weo_a_2050. Hence, I'd say every `clustered ` which is assigned to the `type` weo, should be analysed with the `grouped_by ` value that we get from the scenarios dataset for weo, in this case weo_a_2050. 


Regarding https://github.com/2DegreesInvesting/tiltIndicator/pull/739#issuecomment-2111332672

> Here I reproduce the example from your [GoogleSheet](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064), first case by case then all at once. Please review and try to explain how we can change this output to meet your expectations.
> 
> Case by case the outputs seem to make sense but when taken together the output clearly has more values of grouped_by than the ones you expect. The solution may be to gather consolidate the rows where grouped_by is unmatched_NA_NA and weo_NA_NA. But now I need a sleeping-break so I pass the thinking ball to you :-)

I indeed think that we somehow need to consolidate the `grouped_by` unmatched_NA_NA and weo_NA_NA. While the NA values behind `grouped_by` unmatched_NA_NA should be consolidated into both ipr_a_2050 and weo_a_2050, the NA values behind `grouped_by` weo_NA_NA should only be consolidated into weo_a_2050. The reason for that is: 

- the `clustered b `behind `grouped_by` unmatched_NA_NA has no results for either weo or ipr, i.e. on product level, it should have the risk_category NA for both weo_a_2050 and ipr_a_2050 instead. 
- the `clustered` c behind `grouped_by` weo_NA_NA has results for IPR but not for WEO. Hence, on product-level it should only show NA for the `grouped_by `weo_a_2050 but for ipr_a_2050, it should show the actual risk_category value (as it does right now). 

In essence, I believe we should in the end only have the two `grouped_by ` ipr_a_2050 and weo_a_2050. And every clustered where we have the `type` "weo" should be assigned to `grouped_by` weo_a_2050, while every clustered where we have the `type` "ipr" should be assigned to `grouped_by` ipr_a_2050, irrespective of whether it has a `sector `or `subsector`. If a clustered has no `type` at all (or as denoted in the example a `type` "unmatched"), it should be countred as NA for all type_scenario_year combinations in the scenarios dataset. 

Does that somewhat help?

Tilmon commented:
@maurolepore additional thoughts: 

I was wondering how we can put my thoughts from https://github.com/2DegreesInvesting/tiltIndicator/pull/739#issuecomment-2112020721 in a clear business logic. How about the explanation below? Does that help to make things clearer? It's the same thing as above, just expressed in a different order. 

The dataset scenarios determines the list of values for `grouped_by` that we have. And then every `clustered` has to be assigned to all `grouped_by` values. Means in the following example...

<img width="221" alt="image" src="https://github.com/2DegreesInvesting/tiltIndicator/assets/88719608/95283121-f776-4fc6-a149-0a606ab90c13">

... that the only two possible `grouped_by` values are ipr_a_2050 and weo_a_2050. And now all `clustered` need to be attributed to both values. In this example...

<img width="373" alt="image" src="https://github.com/2DegreesInvesting/tiltIndicator/assets/88719608/38c7cf18-381c-456a-af5a-c1b44da1f5c3">

- ... `clustered` "a" leads to `risk_category` low/medium/high in both `grouped_by`, because it has sectors for ipr & weo. 
- ... `clustered` "b" leads to `risk_category` NA in both `grouped_by`, because it has no sector for neither ipr nor weo
- ... `clustered` "c" leads to `risk_category` low/medium/high in for ipr_a_2050 and to NA for weo_a_2050 because it has a sector only for ipr, not for weo.

Not sure if this really helps. But worth a try I hope :) 

cc' @AnneSchoenauer

Tilmon commented:
Dear @maurolepore I discussed the suggestions I made in the comments above with @AnneSchoenauer and she agrees that that would be the ideal way. If it's not possible, we need to explore alternatives. 
To put my thoughts from https://github.com/2DegreesInvesting/tiltIndicator/pull/739#issuecomment-2112327196 into concrete examples, I created a new tab in the [Google Sheet (_v2)](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=358627689) which should now reflect the "business logic". Please note that I colored all cells green that have changed - to make it easier for you to see the difference. 

Thanks and please reach out if it's unclear!

maurolepore commented:
Thanks @Tilmon, your comments help.

Here I'll answer this question:

> I see in the outputs the column `profile_ranking`. Does this column show the reductions? 

Yes, [for the `sector*()` functions the column `profile_ranking` maps to reductions](https://github.com/2DegreesInvesting/tiltIndicator/blob/fb60f571ce530a4c17316f0f7c1e86f967052d26/R/spa.R#L45).

> Then can we also call it that way?

Maybe in tiltIndicatorAfter, but tiltIndicator would not be a good place for that kind of indicator-specific change. tiltIndicator sits at the core of the system and manipulates business logic at a level that is mostly abstract and general.

The name `profile_ranking` may not be perfect for the `sector*()` functions specifically, but seemed good in that it uses a general concept from tilt's [domain-specific-language](https://en.wikipedia.org/wiki/Domain-specific_language) that applies to all indicators. Such a general-name can then be used to programatically refer to the columns of all the outputs of all the indicators. This standardization makes code dramatically more maintainable. 

For a concrete example, [note how simple was the change in the related PR](https://github.com/2DegreesInvesting/tiltIndicator/pull/613/files). This change propagates through the function `cols_at_product_level()` to many parts of tiltIndicator. 

Similarly we use `cols_at_all_levels()` which also make the code easier to maintain, for example by automatically updating the name of the columns in the [Value section of each indicator's helpfile](https://2degreesinvesting.github.io/tiltIndicator/reference/emissions_profile.html#value). Take a moment to note that those column names are NOT mentioned explicitely in the function that generates that documentation: [`document_default_value()`](https://github.com/2DegreesInvesting/tiltIndicator/blob/fb60f571ce530a4c17316f0f7c1e86f967052d26/R/document_default_value.R#L9)).

If you still think `profile_ranking` is not a good name, we can open a new issue and refer to this comment. But before we go through that trouble, consider that tiltIndicator does not face users. The output of tiltIndicator is consumed by tiltIndicatorAfter, and there is where you need to worry about the user-facing names. 

You can visualize the aspirational architecture of our system using [The Clean Architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) model. tiltIndicator aims to host the enterprise (yellow) and application (red) business rules, and tiltIndicatorAfter aims to host the "interface adaptors" (green). 

![image](https://github.com/2DegreesInvesting/tiltIndicator/assets/5856545/46a8968d-c3a4-4e33-a652-ca4743d47d44)

### The introduction of `profile_ranking`

* [Changelog entry introducing `profile_ranking`](https://2degreesinvesting.github.io/tiltIndicator/news/index.html#tiltindicator-0009102)
* [Related issue](https://github.com/2DegreesInvesting/tiltIndicator/issues/549).

maurolepore commented:
@Tilmon

Here's today's update.

### Good news

The code now yields exactly what you expect in your example [googlesheet v1](https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064).

<details/>
<summary/>
reprex
</summary>

``` r
devtools::load_all()
#> ℹ Loading tiltIndicator

companies <- tribble(
  ~companies_id, ~clustered, ~activity_uuid_product_uuid, ~tilt_sector, ~tilt_subsector,       ~type,     ~sector,  ~subsector,
            "a",        "a",                         "a",          "a",             "a",       "ipr",     "total",    "energy",
            "a",        "a",                         "a",          "a",             "a",       "weo",     "total",    "energy",
            "a",        "b",                 "unmatched",  "unmatched",     "unmatched", "unmatched", "unmatched", "unmatched",
            "a",        "c",                 "unmatched",          "c",             "c",       "ipr", "land use",   "land use",
            "a",        "c",                 "unmatched",          "c",             "c",       "weo",         NA,           NA
)

scenarios <- tribble(
     ~sector,   ~subsector,  ~year, ~reductions, ~type, ~scenario,
     "total",     "energy",   2050,         1.0, "ipr",       "a",
     "total",     "energy",   2050,         0.6, "weo",       "a",
  "land use",   "land use",   2050,         0.3, "ipr",       "a"
)

# ALL AT ONCE ----------------------------------------------------------------

sector_profile(companies, scenarios) |> unnest_product() |> arrange(clustered)
#> # A tibble: 5 × 11
#>   companies_id grouped_by risk_category profile_ranking clustered
#>   <chr>        <chr>      <chr>                   <dbl> <chr>    
#> 1 a            ipr_a_2050 high                      1   a        
#> 2 a            weo_a_2050 medium                    0.6 a        
#> 3 a            <NA>       <NA>                     NA   b        
#> 4 a            ipr_a_2050 low                       0.3 c        
#> 5 a            weo_a_2050 <NA>                     NA   c        
#> # ℹ 6 more variables: activity_uuid_product_uuid <chr>, tilt_sector <chr>,
#> #   scenario <chr>, year <dbl>, type <chr>, tilt_subsector <chr>

sector_profile(companies, scenarios) |> unnest_company()
#> # A tibble: 8 × 4
#>   companies_id grouped_by risk_category value
#>   <chr>        <chr>      <chr>         <dbl>
#> 1 a            ipr_a_2050 high          0.333
#> 2 a            ipr_a_2050 medium        0    
#> 3 a            ipr_a_2050 low           0.333
#> 4 a            ipr_a_2050 <NA>          0.333
#> 5 a            weo_a_2050 high          0    
#> 6 a            weo_a_2050 medium        0.333
#> 7 a            weo_a_2050 low           0    
#> 8 a            weo_a_2050 <NA>          0.667
```

</details>



























### Bad news

Although we're closer, this is not yet the end. 

The same example doesn't work smoothly when each case is considered separately.

<details/>
<summary/>
reprex
</summary>

``` r
devtools::load_all()
#> ℹ Loading tiltIndicator

companies <- tribble(
  ~companies_id, ~clustered, ~activity_uuid_product_uuid, ~tilt_sector, ~tilt_subsector,       ~type,     ~sector,  ~subsector,
            "a",        "a",                         "a",          "a",             "a",       "ipr",     "total",    "energy",
            "a",        "a",                         "a",          "a",             "a",       "weo",     "total",    "energy",
            "a",        "b",                 "unmatched",  "unmatched",     "unmatched", "unmatched", "unmatched", "unmatched",
            "a",        "c",                 "unmatched",          "c",             "c",       "ipr", "land use",   "land use",
            "a",        "c",                 "unmatched",          "c",             "c",       "weo",         NA,           NA
)

scenarios <- tribble(
     ~sector,   ~subsector,  ~year, ~reductions, ~type, ~scenario,
     "total",     "energy",   2050,         1.0, "ipr",       "a",
     "total",     "energy",   2050,         0.6, "weo",       "a",
  "land use",   "land use",   2050,         0.3, "ipr",       "a"
)

# CASE BY CASE ---------------------------------------------------------------

case_a <- filter(companies, clustered == "a")
case_a
#> # A tibble: 2 × 8
#>   companies_id clustered activity_uuid_produc…¹ tilt_sector tilt_subsector type 
#>   <chr>        <chr>     <chr>                  <chr>       <chr>          <chr>
#> 1 a            a         a                      a           a              ipr  
#> 2 a            a         a                      a           a              weo  
#> # ℹ abbreviated name: ¹​activity_uuid_product_uuid
#> # ℹ 2 more variables: sector <chr>, subsector <chr>

sector_profile(case_a, scenarios) |> unnest_product()
#> # A tibble: 2 × 11
#>   companies_id grouped_by risk_category profile_ranking clustered
#>   <chr>        <chr>      <chr>                   <dbl> <chr>    
#> 1 a            ipr_a_2050 high                      1   a        
#> 2 a            weo_a_2050 medium                    0.6 a        
#> # ℹ 6 more variables: activity_uuid_product_uuid <chr>, tilt_sector <chr>,
#> #   scenario <chr>, year <dbl>, type <chr>, tilt_subsector <chr>

sector_profile(case_a, scenarios) |> unnest_company()
#> # A tibble: 8 × 4
#>   companies_id grouped_by risk_category value
#>   <chr>        <chr>      <chr>         <dbl>
#> 1 a            ipr_a_2050 high              1
#> 2 a            ipr_a_2050 medium            0
#> 3 a            ipr_a_2050 low               0
#> 4 a            ipr_a_2050 <NA>              0
#> 5 a            weo_a_2050 high              0
#> 6 a            weo_a_2050 medium            1
#> 7 a            weo_a_2050 low               0
#> 8 a            weo_a_2050 <NA>              0




case_b <- filter(companies, clustered == "b")
case_b
#> # A tibble: 1 × 8
#>   companies_id clustered activity_uuid_produc…¹ tilt_sector tilt_subsector type 
#>   <chr>        <chr>     <chr>                  <chr>       <chr>          <chr>
#> 1 a            b         unmatched              unmatched   unmatched      unma…
#> # ℹ abbreviated name: ¹​activity_uuid_product_uuid
#> # ℹ 2 more variables: sector <chr>, subsector <chr>

sector_profile(case_b, scenarios) |> unnest_product()
#> Error in `dplyr_col_modify()`:
#> ! Can't recycle `grouped_by` (size 2) to size 0.

sector_profile(case_b, scenarios) |> unnest_company()
#> Error in `dplyr_col_modify()`:
#> ! Can't recycle `grouped_by` (size 2) to size 0.




case_c <- filter(companies, clustered == "c")
case_c
#> # A tibble: 2 × 8
#>   companies_id clustered activity_uuid_produc…¹ tilt_sector tilt_subsector type 
#>   <chr>        <chr>     <chr>                  <chr>       <chr>          <chr>
#> 1 a            c         unmatched              c           c              ipr  
#> 2 a            c         unmatched              c           c              weo  
#> # ℹ abbreviated name: ¹​activity_uuid_product_uuid
#> # ℹ 2 more variables: sector <chr>, subsector <chr>

sector_profile(case_c, scenarios) |> unnest_product()
#> # A tibble: 2 × 11
#>   companies_id grouped_by risk_category profile_ranking clustered
#>   <chr>        <chr>      <chr>                   <dbl> <chr>    
#> 1 a            ipr_a_2050 low                       0.3 c        
#> 2 a            weo_a_2050 <NA>                     NA   c        
#> # ℹ 6 more variables: activity_uuid_product_uuid <chr>, tilt_sector <chr>,
#> #   scenario <chr>, year <dbl>, type <chr>, tilt_subsector <chr>

sector_profile(case_c, scenarios) |> unnest_company()
#> # A tibble: 8 × 4
#>   companies_id grouped_by risk_category value
#>   <chr>        <chr>      <chr>         <dbl>
#> 1 a            ipr_a_2050 high              0
#> 2 a            ipr_a_2050 medium            0
#> 3 a            ipr_a_2050 low               1
#> 4 a            ipr_a_2050 <NA>              0
#> 5 a            weo_a_2050 high              0
#> 6 a            weo_a_2050 medium            0
#> 7 a            weo_a_2050 low               0
#> 8 a            weo_a_2050 <NA>              1
```

</details>







Also multiple previous tests fail.

<details/>
<summary/>
reprex
</summary>


``` r
devtools::test_active_file("R/sector_profile.R")
#> 
#> [ FAIL 0 | WARN 0 | SKIP 0 | PASS 0 ]
#> [ FAIL 0 | WARN 0 | SKIP 0 | PASS 1 ]
#> [ FAIL 0 | WARN 0 | SKIP 0 | PASS 2 ]
#> [ FAIL 0 | WARN 0 | SKIP 0 | PASS 3 ]
#> [ FAIL 0 | WARN 0 | SKIP 0 | PASS 4 ]
#> [ FAIL 1 | WARN 0 | SKIP 0 | PASS 4 ]
#> [ FAIL 2 | WARN 0 | SKIP 0 | PASS 4 ]
#> [ FAIL 3 | WARN 0 | SKIP 0 | PASS 4 ]
#> [ FAIL 4 | WARN 0 | SKIP 0 | PASS 4 ]
#> [ FAIL 4 | WARN 0 | SKIP 0 | PASS 5 ]
#> [ FAIL 4 | WARN 0 | SKIP 0 | PASS 6 ]
#> [ FAIL 4 | WARN 0 | SKIP 0 | PASS 7 ]
#> [ FAIL 5 | WARN 0 | SKIP 0 | PASS 7 ]
#> [ FAIL 6 | WARN 0 | SKIP 0 | PASS 7 ]
#> [ FAIL 7 | WARN 0 | SKIP 0 | PASS 7 ]
#> [ FAIL 8 | WARN 0 | SKIP 0 | PASS 7 ]
#> [ FAIL 8 | WARN 0 | SKIP 0 | PASS 8 ]
#> [ FAIL 8 | WARN 0 | SKIP 0 | PASS 9 ]
#> [ FAIL 8 | WARN 0 | SKIP 0 | PASS 10 ]
#> [ FAIL 8 | WARN 0 | SKIP 0 | PASS 11 ]
#> 
#> ── Failure ('test-sector_profile.R:44:3'): at product level, preserves unmatched products ──
#> "unmatched" %in% out[[aka("uid")]] is not TRUE
#> 
#> `actual`:   FALSE
#> `expected`: TRUE 
#> 
#> ── Failure ('test-sector_profile.R:58:3'): at product level, unmatched product yield `NA` in the expected columns ──
#> is.na(out$grouped_by) is not TRUE
#> 
#> `actual`:       
#> `expected`: TRUE
#> 
#> ── Failure ('test-sector_profile.R:59:3'): at product level, unmatched product yield `NA` in the expected columns ──
#> is.na(out$risk_category) is not TRUE
#> 
#> `actual`:       
#> `expected`: TRUE
#> 
#> ── Failure ('test-sector_profile.R:60:3'): at product level, unmatched product yield `NA` in the expected columns ──
#> is.na(out$profile_ranking) is not TRUE
#> 
#> `actual`:       
#> `expected`: TRUE
#> 
#> ── Failure ('test-sector_profile.R:108:3'): at company level, one matched and one unmatched products yield `value = 1/2` where `risk_category = NA` and in one other `risk_category` (#657) ──
#> `na` (`actual`) not equal to 1/2 (`expected`).
#> 
#>   `actual`: 0.0
#> `expected`: 0.5
#> 
#> ── Failure ('test-sector_profile.R:110:3'): at company level, one matched and one unmatched products yield `value = 1/2` where `risk_category = NA` and in one other `risk_category` (#657) ──
#> sort(other) (`actual`) not equal to c(0, 0, 1/2) (`expected`).
#> 
#>   `actual`: 0.0 0.0 1.0
#> `expected`: 0.0 0.0 0.5
#> 
#> ── Failure ('test-sector_profile.R:126:3'): at company level, two matched and one unmatched products yield `value = 1/3` where `risk_category = NA` and `value = 2/3` in one other `risk_category` (#657) ──
#> `na` (`actual`) not equal to 1/3 (`expected`).
#> 
#>   `actual`: 0.0
#> `expected`: 0.3
#> 
#> ── Failure ('test-sector_profile.R:128:3'): at company level, two matched and one unmatched products yield `value = 1/3` where `risk_category = NA` and `value = 2/3` in one other `risk_category` (#657) ──
#> sort(other) (`actual`) not equal to c(0, 0, 2/3) (`expected`).
#> 
#>   `actual`: 0.0 0.0 1.0
#> `expected`: 0.0 0.0 0.7
#> 
#> [ FAIL 8 | WARN 0 | SKIP 0 | PASS 11 ]
```

</details>

So I need to investigate each problematic result. In some cases the solution may be obvious, and the new requirement will lead me to adapt or remove a previous test. In other cases the conflict between the old and new requirements may not be obvious and we'll need to discuss further to decide what the code should actually do.

I'll come back tomorrow with a fresh brain.

Tilmon commented:
@maurolepore that's indeed good news! Awesome. 

I hope the Bad News won't cause too much headache to solve. Good luck!

maurolepore commented:
@Tilmon, 

`sector_profile()` now works as you expect. I fixed all tests and added a few more. 

`sector_profile_upstream()` didn't get any update yet. Can you please help me by modifying this draft example so that I have something to test against? 

It's bases on your googlesheet example for `sector_profile()` but it's a bit more complex since we need an additional `inputs` dataset.

<details/>
<summary/>
reprex
</summary>

``` r
devtools::load_all()
#> ℹ Loading tiltIndicator

# styler: off
companies <- tribble(
~companies_id, ~clustered, ~activity_uuid_product_uuid, ~tilt_sector, ~tilt_subsector,
          "a",        "a",                         "a",          "a",             "a",
          "a",        "a",                         "a",          "a",             "a",
          "a",        "b",                 "unmatched",  "unmatched",     "unmatched",
          "a",        "c",                 "unmatched",          "c",             "c",
          "a",        "c",                 "unmatched",          "c",             "c"
)

inputs <- tribble(
  ~activity_uuid_product_uuid, ~input_tilt_sector, ~input_tilt_subsector,       ~type,     ~sector,  ~subsector, ~input_activity_uuid_product_uuid,
                          "a",                "a",                   "a",       "ipr",     "total",    "energy",                               "a",
                          "a",                "a",                   "a",       "weo",     "total",    "energy",                               "a",
                  "unmatched",        "unmatched",           "unmatched", "unmatched", "unmatched", "unmatched",                       "unmatched",
                  "unmatched",                "c",                   "c",       "ipr",  "land use",  "land use",                       "unmatched",
                  "unmatched",                "c",                   "c",       "weo",          NA,          NA,                       "unmatched"
)

scenarios <- tribble(
     ~sector,   ~subsector,  ~year, ~reductions, ~type, ~scenario,
     "total",     "energy",   2050,         1.0, "ipr",       "a",
     "total",     "energy",   2050,         0.6, "weo",       "a",
  "land use",   "land use",   2050,         0.3, "ipr",       "a"
)
# styler: on

sector_profile_upstream(companies, scenarios, inputs) |> unnest_product()
#> # A tibble: 8 × 13
#>   companies_id grouped_by risk_category profile_ranking clustered
#>   <chr>        <chr>      <chr>                   <dbl> <chr>    
#> 1 a            ipr_a_2050 high                      1   a        
#> 2 a            weo_a_2050 medium                    0.6 a        
#> 3 a            <NA>       <NA>                     NA   b        
#> 4 a            ipr_a_2050 low                       0.3 b        
#> 5 a            <NA>       <NA>                     NA   b        
#> 6 a            <NA>       <NA>                     NA   c        
#> 7 a            ipr_a_2050 low                       0.3 c        
#> 8 a            <NA>       <NA>                     NA   c        
#> # ℹ 8 more variables: activity_uuid_product_uuid <chr>, tilt_sector <chr>,
#> #   scenario <chr>, year <dbl>, type <chr>,
#> #   input_activity_uuid_product_uuid <chr>, input_tilt_sector <chr>,
#> #   input_tilt_subsector <chr>
sector_profile_upstream(companies, scenarios, inputs) |> unnest_product()
#> # A tibble: 8 × 13
#>   companies_id grouped_by risk_category profile_ranking clustered
#>   <chr>        <chr>      <chr>                   <dbl> <chr>    
#> 1 a            ipr_a_2050 high                      1   a        
#> 2 a            weo_a_2050 medium                    0.6 a        
#> 3 a            <NA>       <NA>                     NA   b        
#> 4 a            ipr_a_2050 low                       0.3 b        
#> 5 a            <NA>       <NA>                     NA   b        
#> 6 a            <NA>       <NA>                     NA   c        
#> 7 a            ipr_a_2050 low                       0.3 c        
#> 8 a            <NA>       <NA>                     NA   c        
#> # ℹ 8 more variables: activity_uuid_product_uuid <chr>, tilt_sector <chr>,
#> #   scenario <chr>, year <dbl>, type <chr>,
#> #   input_activity_uuid_product_uuid <chr>, input_tilt_sector <chr>,
#> #   input_tilt_subsector <chr>
```

</details>

Tilmon commented:
Hi @maurolepore , awesome that the `sector_profile()` now shows the results as expected!! Congrats! 

Re `sector_upstream_profile()`: @AnneSchoenauer and I discussed priorities and decided that given our ambitious target to launch the webtool in June, it's best to pause the `sector_upstream_profile()` work for now. It's anyways a bit tricky to publish the licensed input data in the webtool. Therefore, we suggest to launch the webtool at least in the first version without the upstream results. 

Would therefore be great if you could instead soon start working on the webtool!

kalashsinghal commented on tests/testthat/test-sector_profile.R:
```diff
@@ -127,3 +134,257 @@ test_that("at company level, two matched and one unmatched products yield `value
+
+# TODO: Create a constructor for Tilman's example
+
+test_that("at product level, Tilman's example yields what he expects", {
+  # https://docs.google.com/spreadsheets/d/16u9WNtVY-yDsq6kHANK3dyYGXTbNQ_Bn/edit#gid=156243064
```
For review, it would be great if we know what's special about Tilman's example. right?

kalashsinghal commented on R/sector_profile_at_product_level.R:
```diff
@@ -32,3 +34,54 @@ sp_cols_at_product_level <- function() {
+    filter(!is.na(.data$grouped_by)) |>
+    group_by(.data[[aka("id")]], .data[[aka("cluster")]]) |>
+    mutate(grouped_by = factor(.data$grouped_by, levels = levels)) |>
+    expand(.data$grouped_by) |>
+    left_join(data, by = c(aka("id"), aka("cluster"), "grouped_by")) |>
```
@maurolepore As suggested by you, don't you think you can avoid hard-coding brittle column names like `by = "grouped_by"`?

maurolepore replied:
Thanks @kalashsinghal :-)

Yes, I need to cleanup this PR a bit before merging and indeed it would be more maintainable to hide "grouped_by" behind an interface. But in general I'm comfortable using the literal "grouped_by" than "companies_id" or "clustered". 

The reason is that the name "grouped_by" is fully under the control of this package. It's a name in the data structure ("entity") that this package defines to work conveniently with data across all indicators. As such it's way more stable than for example "companies_id" or "clustered" -- which are controlled by external agencies (our managers, the maintainers of tiltIndicatorBefore, etc.).

kalashsinghal reviewed (approved):
LGTM! Please address my comments before merging! :)

maurolepore commented:
@Tilmon, noted. Thanks.

https://github.com/2DegreesInvesting/tiltIndicator/issues/784
